Implement declarative (`macro_rules!`) attribute macros (RFC 3697)

joshtriplett opened the pull request:
This implements [RFC 3697](https://github.com/rust-lang/rust/issues/143547), "Declarative (`macro_rules!`) attribute macros".

I would suggest reading this commit-by-commit. This first introduces the
feature gate, then adds parsing for attribute rules (doing nothing with them),
then adds the ability to look up and apply `macro_rules!` attributes by path,
then adds support for local attributes, then adds a test, and finally makes
various improvements to errors.

---

- *See also the related https://github.com/rust-lang/rust/pull/145208*

joshtriplett committed:
mbe: Parse macro attribute rules

This handles various kinds of errors, but does not allow applying the
attributes yet.

This adds the feature gate `macro_attr`.

joshtriplett committed:
mbe: Handle applying attribute rules with paths

Add infrastructure to apply an attribute macro given argument tokens and
body tokens.

Teach the resolver to consider `macro_rules` macros when looking for an
attribute via a path.

This does not yet handle local `macro_rules` attributes.

joshtriplett committed:
mbe: Handle local `macro_rules` attr resolution

Teach the resolver to consider `macro_rules` macros when looking for a
local attribute. When looking for an attribute and considering a
`macro_rules` macro, load the macro in order to see if it has attribute
rules.

Include a FIXME about tracking multiple macro kinds for a Def instead.

joshtriplett committed:
mbe: Add test for `macro_rules` attributes

Test macros via path and local macros.

joshtriplett committed:
mbe: In error messages, don't assume attributes are always proc macros

Now that `macro_rules` macros can define attribute rules, make sure
error messages account for that.

joshtriplett committed:
mbe: Fix error message for using a macro with no `attr` rules as an attribute

Avoid saying "a declarative macro cannot be used as an attribute macro";
instead, say that the macro has no `attr` rules.

Veykril commented on compiler/rustc_errors/src/emitter.rs:
```diff
@@ -417,7 +417,7 @@ pub trait Emitter {
                     let msg: Cow<'static, _> = match trace.kind {
                         ExpnKind::Macro(MacroKind::Attr, _) => {
-                            "this procedural macro expansion".into()
+                            "this macro attribute expansion".into()
```
I'd say for consistency this should either state `this attribute macro expansion` or for derives below to be changed to `this macro derive expansion`

theemathas commented:
Here's an edge case that probably needs a test.

<details>
<summary>Code and error</summary>

```rust
#![feature(macro_attr)]

#[foo]
macro_rules! foo {
    attr() { $($body:tt)* } => {
        $($body)*
    }
}
use foo;
```

```
warning: unused macro definition: `foo`
 --> src/lib.rs:4:14
  |
4 | macro_rules! foo {
  |              ^^^
  |
  = note: `#[warn(unused_macros)]` on by default

error: cannot determine resolution for the attribute macro `foo`
 --> src/lib.rs:3:3
  |
3 | #[foo]
  |   ^^^
  |
  = note: import resolution is stuck, try simplifying macro imports

warning: `foo` (lib) generated 1 warning
error: could not compile `foo` (lib) due to 1 previous error; 1 warning emitted
```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16579623311/job/46892403859) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46892403859) [(plain)](https://github.com/rust-lang/rust/commit/82d889045a6a49dc77d2e986e083608688a7de48/checks/46892403859/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
error: using `.clone()` on a ref-counted pointer
   --> compiler/rustc_expand/src/mbe/macro_rules.rs:634:78
    |
634 |         has_attr_rules.then(|| Arc::new(mk_syn_ext(SyntaxExtensionKind::Attr(expander.clone()))));
    |                                                                              ^^^^^^^^^^^^^^^^ help: try: `Arc::<mbe::macro_rules::MacroRulesMacroExpander>::clone(&expander)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr
    = note: requested on the command line with `-D clippy::clone-on-ref-ptr`

[RUSTC-TIMING] rustc_expand test:false 2.357
```

</details>

joshtriplett commented:
@theemathas I can't reproduce that here. I get a different error instead:

```text
error: cannot find attribute `foo` in this scope
  --> $DIR/macro-rules-attr-cycle.rs:4:3
   |
LL | #[foo]
   |   ^^^
LL | macro_rules! foo {
   |              --- `foo` exists, but has no `attr` rules
```

This isn't the right error message, but the error is getting caught.

(Suggestions welcome for how to catch and handle this case better.)

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16580911961/job/46896587693) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46896587693) [(plain)](https://github.com/rust-lang/rust/commit/d94b305bc80b047e7d80d8f6646e7fba9e5e1b9e/checks/46896587693/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_incremental v0.0.0 (/checkout/compiler/rustc_incremental)
error[E0308]: mismatched types
   --> compiler/rustc_expand/src/mbe/macro_rules.rs:633:89
    |
633 |         has_attr_rules.then(|| Arc::new(mk_syn_ext(SyntaxExtensionKind::Attr(Arc::clone(exp)))));
    |                                                                              ---------- ^^^ expected `&Arc<...>`, found `Arc<MacroRulesMacroExpander>`
    |                                                                              |
    |                                                                              arguments to this function are incorrect
    |
    = note: expected reference `&Arc<dyn base::AttrProcMacro + DynSend + DynSync>`
                  found struct `Arc<MacroRulesMacroExpander>`
note: method defined here
   --> /rustc/390a0ab5dc4a895235a551e502c3893c3337731d/library/core/src/clone.rs:204:8

For more information about this error, try `rustc --explain E0308`.
[RUSTC-TIMING] rustc_expand test:false 1.077
error: could not compile `rustc_expand` (lib) due to 1 previous error
```

</details>

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -380,6 +464,60 @@ pub(super) fn try_match_macro<'matcher, T: Tracker<'matcher>>(
+/// named_matches if it was successful, and nothing if it failed. On failure, it's the caller's job
+/// to use `track` accordingly to record all errors correctly.
+#[instrument(level = "debug", skip(psess, attr_args, attr_body, rules, track), fields(tracking = %T::description()))]
+pub(super) fn try_match_macro_attr<'matcher, T: Tracker<'matcher>>(
+    psess: &ParseSess,
+    name: Ident,
+    attr_args: &TokenStream,
+    attr_body: &TokenStream,
+    rules: &'matcher [MacroRule],
+    track: &mut T,
+) -> Result<(usize, &'matcher MacroRule, NamedMatches), CanRetry> {
```
Tried it a second time, both by attempting to merge the two functions, and by attempting to factor out a helper from the two. In both cases the result looked like spaghetti, because the differences are entwined with control flow, pattern matching, and variable bindings. I don't think it's worth attempting to unify those two.

joshtriplett committed:
mbe: Add a test confirming that a macro attribute can apply itself recursively

This allows a macro attribute to implement default arguments by
reapplying itself with the defaults filled in, for instance.

joshtriplett committed:
mbe: Add a test checking for infinite recursion in macro attributes

joshtriplett commented:
@theemathas I looked into your test case more. I managed to find a way to reproduce the error you got; it only reproduces if there aren't other errors. So, I can reproduce it with the following test:

```rust
#![crate_type = "lib"]
#![feature(macro_attr)]

#[attr]
macro_rules! attr {
    attr() {} => {}
}

#[attr]
struct S;
```

This gives:

```text
error: cannot find attribute `attr` in this scope
  --> /home/josh/src/rust/tests/ui/macros/macro-rules-attr-apply-to-self.rs:4:3
   |
LL | #[attr]
   |   ^^^^
LL | macro_rules! attr {
   |              ---- `attr` exists, but has no `attr` rules

error: cannot determine resolution for the attribute macro `attr`
  --> /home/josh/src/rust/tests/ui/macros/macro-rules-attr-apply-to-self.rs:9:3
   |
LL | #[attr]
   |   ^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports
```

The second error seems reasonable for attempting to invoke an attribute on itself.

The first error also largely seems fine, except that it shouldn't say ```has no `attr` rules``` when the real problem is a cycle. 

Note that "cannot find" is the same error that you get if you try this:

```rust
mac! {
    macro_rules! mac {
        {} => {}
    }
}

mac! {
    struct S;
}
```

This gives:

```text
error: cannot find macro `mac` in this scope
 --> src/main.rs:1:1
  |
1 | mac! {
  | ^^^

error: cannot find macro `mac` in this scope
 --> src/main.rs:7:1
  |
7 | mac! {
  | ^^^
```

So, I think "cannot find ... in this scope" is an acceptable error, albeit one we *might* be able to improve. I think the main problem here is the spurious ``` `attr` exists, but has no `attr` rules```.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16581838706/job/46899478036) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46899478036) [(plain)](https://github.com/rust-lang/rust/commit/0eb49e1b1d1c1bf15e3199b3826cf03d8485326f/checks/46899478036/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_public test:false 5.657
error: using `.clone()` on a ref-counted pointer
   --> compiler/rustc_resolve/src/macros.rs:838:85
    |
838 |                 Some(MacroKind::Attr) if let Some(ref ext) = macro_data.attr_ext => ext.clone(),
    |                                                                                     ^^^^^^^^^^^ help: try: `Arc::<rustc_expand::base::SyntaxExtension>::clone(&ext)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr
    = note: requested on the command line with `-D clippy::clone-on-ref-ptr`

error: using `.clone()` on a ref-counted pointer
   --> compiler/rustc_resolve/src/macros.rs:839:22
    |
839 |                 _ => macro_data.ext.clone(),
    |                      ^^^^^^^^^^^^^^^^^^^^^^ help: try: `Arc::<rustc_expand::base::SyntaxExtension>::clone(&macro_data.ext)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr

[RUSTC-TIMING] rustc_resolve test:false 8.357
error: could not compile `rustc_resolve` (lib) due to 2 previous errors
```

</details>

petrochenkov commented:
(I started looking today, but will continue only on ~Monday.)

petrochenkov commented on tests/ui/macros/macro-rules-attr.rs:
```diff
@@ -0,0 +1,90 @@
+//@ run-pass
+//@ check-run-results
+#![feature(macro_attr)]
+#![warn(unused)]
```
IIRC, this is the default for run-pass tests.

petrochenkov commented on tests/ui/macros/macro-rules-attr.stderr:
```diff
@@ -0,0 +1,21 @@
+  --> $DIR/macro-rules-attr.rs:43:9
+   |
+LL |     attr() {} => {
+   |         ^^^^^
```
Weird span.

joshtriplett replied:
This span is used exclusively for unused rules, for which bang-style macros use the span of the matcher as well. An attr rule being used depends on both the arguments and the body, so this points to both.

That said, this should have used a `MultiSpan`; I just hadn't gone to the trouble to switch to one. I've now done so.

petrochenkov commented on tests/ui/macros/macro-rules-attr-error.stderr:
```diff
@@ -0,0 +1,22 @@
+   |
+   = note: this error originates in the attribute macro `local_attr` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: invoked macro has no invocation rules
```
Strange wording, the macro has rules, just not fn-like rules.

A bit of terminology (from the compiler, didn't check the reference):
- Macro invocation - invocation of any macro - fn-like, or attr, or derive
- Macro call - invocation of a fn-like macro

joshtriplett replied:
The reference seemed to use the term "invoke" specifically for bang invocation, not attributes or derives. I can come up with another way to describe this, though.

petrochenkov commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -1179,6 +1182,9 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
 
         if let Some(builtin_name) = ext.builtin_name {
+            if attr_ext.is_some() {
+                self.dcx().emit_err(errors::BuiltinMacroWithNonInvocationRules { span, ident });
```
I don't think this is necessary, `#[rustc_builtin_macro]` is unstable, and any rules of such macros are ignored anyway.

joshtriplett replied:
@petrochenkov Should I just silently set the attr_ext to `None` for a built-in macro, then?

joshtriplett replied:
I've now done this, and deleted the error.

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -371,6 +465,60 @@ pub(super) fn try_match_macro<'matcher, T: Tracker<'matcher>>(
+/// named_matches if it was successful, and nothing if it failed. On failure, it's the caller's job
+/// to use `track` accordingly to record all errors correctly.
+#[instrument(level = "debug", skip(psess, attr_args, attr_body, rules, track), fields(tracking = %T::description()))]
+pub(super) fn try_match_macro_attr<'matcher, T: Tracker<'matcher>>(
```
They are similar, but attempting to merge them produces spaghetti. It's theoretically possible but the result seems substantially worse than the current state.

petrochenkov replied:
Ok.

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -468,6 +650,19 @@ fn check_no_eof(sess: &Session, p: &Parser<'_>, msg: &'static str) -> Option<Err
+    if let tokenstream::TokenTree::Delimited(dspan, _, delim, _) = args
+        && *delim != Delimiter::Parenthesis
+    {
+        return Err(sess.dcx().emit_err(errors::MacroArgsBadDelim {
```
Done, including a test.

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -400,9 +548,31 @@ pub fn compile_declarative_macro(
+            if !features.macro_attr() {
+                let msg = "`macro_rules!` attributes are unstable";
+                let e = feature_err(sess, sym::macro_attr, span, msg);
+                return dummy_syn_ext(e.emit());
```
Done.

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -400,9 +548,31 @@ pub fn compile_declarative_macro(
+            }
+            let args = p.parse_token_tree();
+            check_emission(check_args_parens(sess, &args));
+            let args = parse_one_tt(args, RulePart::Pattern, sess, node_id, features, edition);
```
@petrochenkov I'm not sure what you're proposing to merge here. The only common logic adjacent to this line would be the call to `check_lhs` which is already factored out into a function.

joshtriplett commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -122,10 +125,11 @@ impl<'a> ParserAnyMacro<'a> {
+    /// A function-style rule, for use with `m!()`
+    Func { lhs: Vec<MatcherLoc>, lhs_span: Span, rhs: mbe::TokenTree },
+    /// An attr rule, for use with `#[m]`
+    Attr { args: Vec<MatcherLoc>, body: Vec<MatcherLoc>, lhs_span: Span, rhs: mbe::TokenTree },
```
@petrochenkov Using an enum here ensures that every user of this enum has to be updated to handle each variant. Every user of it needs distinct logic for these two cases. If they (for instance) used `args.is_none()` or `args.is_some()` instead, that would make it possible to access one of the fields without doing that check. An enum ensures you *have* to check the variant before using a field of a rule.

petrochenkov replied:
Ok.

joshtriplett commented on compiler/rustc_expand/src/mbe/diagnostics.rs:
```diff
@@ -47,15 +58,22 @@ pub(super) fn failed_to_match_macro(
+        if attr_args.is_none() && !rules.iter().any(|rule| matches!(rule, MacroRule::Func { .. })) {
+            let mut err = psess.dcx().struct_span_err(sp, "invoked macro has no invocation rules");
+            if !def_head_span.is_dummy() {
+                err.span_label(def_head_span, "this macro has no rules to invoke");
```
I would love to report it early, but knowing whether a macro has invocation rules or not requires parsing it, and as far as I can tell, the compiler is trying to defer parsing macros until it has to. Reporting this earlier would force parsing macros earlier than the compiler currently does.

Could we defer that, and try it in a separate PR?

petrochenkov replied:
>as far as I can tell, the compiler is trying to defer parsing macros until it has to

That's because macro definitions were historically parsed as macro invocations, as you could notice in https://github.com/rust-lang/rust/pull/143070.
In the end all of them end up parsed anyway (unless they are configured away), even if they are never called.

Ideally macros would be parsed as any other items, and stored in AST in a structured form, and then stored into metadata in a structured "pre-compiled" form.
In particular, this would allow use to avoid re-compiling all used macros coming from other crates in every crate.

This is a slightly breaking change though, if done naively, because currently syntactically incorrect macros are accepted under `cfg(false)`.

joshtriplett replied:
Alright. I'm currently working on switching macro kinds over to being bitflags, which is a major change but seems like it's working. When that's done, I'll take a look at trying to handle these errors in a more uniform way.

petrochenkov replied:
Could you add a FIXME here saying that we need to report this at name resolution time, and not at expansion time, using something like `MacroDefKind`?

joshtriplett commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -625,9 +625,20 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
+                                && macro_kind == Some(MacroKind::Attr)
+                                && this
+                                    .get_macro(binding.res())
+                                    .is_some_and(|macro_data| macro_data.attr_ext.is_some())))
```
I'll give this a shot. I'm expecting it to be fairly invasive, but let's see.

joshtriplett replied:
@petrochenkov I'm most of the way done with that change (adding a `MacroKinds` bitflags type, separate from `MacroKind`). But if you're saying that can come in a follow-up PR, that's great; I'd be glad to keep it separate. :)

petrochenkov replied:
Could you add a FIXME here saying that we need to replace this with a `MacroDefKind` supporting multiple call kinds?

bors commented:
:umbrella: The latest upstream changes (presumably #144997) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett committed:
mbe: Add parser test for macro attribute recovery

joshtriplett committed:
mbe: Add test for attribute expansion with `compile_error!`

joshtriplett committed:
mbe: Emit an error if a macro call has no function-like rules

Add a FIXME for moving this error earlier.

joshtriplett committed:
mbe: Add a test for calling a macro with no function-like rules

bors commented:
:pushpin: Commit f88839d2abc1c190c3072638518ff06b79686de1 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov f88839d2abc1c190c3072638518ff06b79686de1 -->
<!-- homu: {"type":"Approved","sha":"f88839d2abc1c190c3072638518ff06b79686de1","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/144787#issuecomment-3168765168) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

joshtriplett commented:
@petrochenkov https://github.com/rust-lang/rust/pull/145153 implements the change to support a macro with multiple kinds. Turned out very nicely, and led naturally to fixing and cleaning up various things. Thanks!
